azkaban-exec-server/src/main/java/azkaban/dag/Node.java

```diff
@@ -0,0 +1,196 @@
+    return this.flow;
+  }
+
+  public void setFlow(final Flow flow) {
+    this.flow = flow;
+  }
+
+  private Node addParent(final Node node) {
+    this.parents.add(node);
+    return this;
+  }
+
+  Node addChild(final Node node) {
+    this.children.add(node);
+    node.addParent(this);
+    return this;
+  }
+
+  Node addChildren(final Node... nodes) {
+    for (final Node node : nodes) {
+      addChild(node);
+    }
+    return this;
+  }
+
+  boolean hasParent() {
+    return !this.parents.isEmpty();
+  }
+
+  boolean isReady() {
+    if (this.status != Status.READY) {
+      return false;
+    }
+    for (final Node parent : this.parents) {
+      if (!parent.isSuccessEffectively()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public void run() {
+    changeStatus(Status.RUNNING);
+  }
+
+  void markSuccess() {
+    changeStatus(Status.SUCCESS);
+    for (final Node child : this.children) {
+      child.check();
+    }
+    this.flow.checkFinished();
+  }
+
+  /**
+   * Checks if all the dependencies are met and run if they are.
+   */
+  private void check() {
+    if (this.status == Status.DISABLED) {
+      return;
+    }
+
+    for (final Node node : this.parents) {
+      if (!node.isSuccessEffectively()) {
+        return;
+      }
+    }
+
+    run();
+  }
+
+  /**
+   * @return true if the dependency requirement is considered satisfied
+   */
+  private boolean isSuccessEffectively() {
+    return Status.isSuccessEffectively(this.status);
+  }
+
+  void markFailure() {
+    changeStatus(Status.FAILURE);
+    for (final Node child : this.children) {
+      child.cancel();
+    }
+    this.flow.checkFinished();
+  }
+
+  private void cancel() {
+    // The node can't be in the running or killed state since one of its ancestors has failed or
+    // been killed. It shouldn't have started.
+    if (this.status != Status.DISABLED) {
+      changeStatus(Status.CANCELED);
+    }
+    for (final Node node : this.children) {
+      node.cancel();
+    }
+  }
+
+  private void changeStatus(final Status status) {
+    this.status = status;
+    this.nodeProcessor.changeStatus(this, this.status);
+  }
+
+  /**
+   * Kills a node.
+   *
+   * Unlike other events, this method expects that the caller will check if the flow is finished.
+   * This action will only be invoked by the {@link Flow#kill()} method, i.e. there is no
+   * DagService#killNode method. In the interest of efficiency, it only needs to check once in
+   * the Flow#kill method.
+   */
+  void kill() {
+    if (this.status == Status.READY || this.status == Status.BLOCKED) {
+      changeStatus(Status.CANCELED);
+    } else if (this.status == Status.RUNNING) {
+      // kill the job
+      changeStatus(Status.KILLING);
+    }
+  }
+
+  void markKilled() {
+    assert (this.status == Status.KILLING);
+    changeStatus(Status.KILLED);
+    this.flow.checkFinished();
+  }
+
+  boolean isInTerminalState() {
+    return Status.isTerminal(this.status);
```

Comment:
Could these methods be simplified to be on the enum instances themselves? For example this line would become:

```java
this.status.isTerminal();
```

Maybe that would even mean that there's no need for this method (`Node#isInTerminalState`).

This has always bothered me with the static methods of `azkaban.executor.Status`: `isStatusRunning` and `isStatusFinished`..

Reply:
Let me try your suggestion.

Reply:
I like your suggestion. Changed.